source/backend/api/Services/DocumentSyncService.cs

```diff
@@ -94,113 +106,122 @@ public ExternalBatchResult SyncMayanMetadataTypes(SyncModel model)
+                var matchingTypeFromMayan = mayanDocumentTypes.Payload.Results.FirstOrDefault(x => x.Id == pimsDocumentTyp.MayanId);
+                if (matchingTypeFromMayan == null)
                 {
-                    var newPimsDocType = new PimsDocumentTyp() { MayanId = mayanDocumentType.Id, DocumentType = mayanDocumentType.Label, DisplayOrder = matchingTypeFromModel?.DisplayOrder };
-                    createdDocumentTypes.Add(documentTypeRepository.Add(newPimsDocType));
+                    createResponseTasks.Add(new AddDocumentToMayanWithNameResponseModel() { AddDocumentTypeTask = mayanDocumentRepository.TryCreateDocumentTypeAsync(new DocumentType() { Label = pimsDocumentTyp.DocumentTypeDescription }), Name = pimsDocumentTyp.DocumentType });
                 }
-                else if (pimsDocumentTypes.FirstOrDefault(x => x.MayanId == mayanDocumentType.Id && (mayanDocumentType.Label != x.DocumentType || x.DisplayOrder != matchingTypeFromModel?.DisplayOrder)) != null)
+                else if (matchingTypeFromMayan.Label != pimsDocumentTyp.DocumentTypeDescription)
                 {
                     // if the Mayan id is the same but the label or display order has changed, update the document type in PIMS.
-                    var updatedPimsDocType = new PimsDocumentTyp() { MayanId = mayanDocumentType.Id, DocumentType = mayanDocumentType.Label, DisplayOrder = matchingTypeFromModel?.DisplayOrder };
-                    updatedDocumentTypes.Add(documentTypeRepository.Update(updatedPimsDocType));
+                    matchingTypeFromMayan.Label = pimsDocumentTyp.DocumentTypeDescription;
+                    updateTasks.Add(mayanDocumentRepository.TryUpdateDocumentTypeAsync(matchingTypeFromMayan));
                 }
             }
+            Task.WaitAll(createResponseTasks.Select(x => x.AddDocumentTypeTask).Concat(updateTasks).ToArray());
+            foreach (var task in updateTasks)
+            {
+                batchResult.UpdatedDocumentType.Add(task.Result);
+            }
+            foreach (var task in createResponseTasks)
+            {
+                if (task != null && task.AddDocumentTypeTask.Result.HttpStatusCode == System.Net.HttpStatusCode.Created)
+                {
+                    var matchingPimsType = pimsDocumentTypes.FirstOrDefault(x => x.DocumentType == task.Name);
+                    matchingPimsType.MayanId = task.AddDocumentTypeTask.Result.Payload.Id;
+                    documentTypeRepository.Update(matchingPimsType);
+                }
+                batchResult.CreatedDocumentType.Add(task.AddDocumentTypeTask.Result);
+            }
+            documentTypeRepository.CommitTransaction();
 
-            IList<PimsDocumentTyp> removedDocumentTypes = new List<PimsDocumentTyp>();
+            pimsDocumentTypes = documentTypeRepository.GetAll(); // re-retrieve the document list from pims as it may have been updated.
```

Comment:
q: why is this necessary? if it has changed the previous steps might be wrong too.